### DISCUSSION TITLE
Stale check runs

### DISCUSSION THREAD

**Discussion by danthedaniel** [community/community#24761] [Apps, API and Webhooks]
<p>I’m planning to create an app that uses the Check Runs API. The checks performed on the code will not be automated, and thus could take a very long time to complete. I noticed that if a check is in progress for more than 14 days GitHub will automatically mark the run as stale (<a href="https://developer.github.com/v3/checks/runs/" rel="nofollow noopener">https://developer.github.com/v3/checks/runs/</a>). I have a couple of questions about stale check runs:</p>
<ol>
<li>Can I still update the run status once it is marked as stale?</li>
<li>If the check is required for merging into a protected branch, will the check be still be required after it is marked as stale?</li>
</ol>

**Comment by francisfuzz** [comment]
<p>:wave: hello there <a class="mention" href="/danthedaniel">@danthedaniel</a>, and welcome to the GitHub Support Community! :tada:</p>
<blockquote>
<ol>
<li>Can I still update the run status once it is marked as stale?</li>
</ol>
</blockquote>
<p>Once a check run transitions from the <code>in_progress</code> state to <code>completed</code> state, it’s not possible to modify its status nor its conclusion further. Also, only GitHub can change a check run conclusion to <code>stale</code> as documented in <a href="https://docs.github.com/en/rest/reference/checks#create-a-check-run--parameters">the parameters section for creating a check run</a>.</p>
<blockquote>
<ol start="2">
<li>If the check is required for merging into a protected branch, will the check be still be required after it is marked as stale?</li>
</ol>
</blockquote>
<p>Yes. If a check is required as a part of a protected branch’s settings, the check must pass as noted in <a href="https://docs.github.com/en/github/administering-a-repository/about-required-status-checks#about-required-status-checks"><code>About required status checks</code></a>. A check run with a <code>stale</code> conclusion is not the same as a check run with a <code>success</code> conclusion. I hope this helps!</p>

**Reply by gaspar09** [reply]
This answer is no longer true based on https://github.com/orgs/community/discussions/24761#discussioncomment-3245367

**Comment by danthedaniel** [comment]
<aside class="quote group-github-employees" data-username="francisfuzz" data-post="2" data-topic="120566">
<div class="title">
<div class="quote-controls"></div>
 francisfuzz:</div>
<blockquote>
<p>Once a check run transitions from the <code>in_progress</code> state to <code>completed</code> state, it’s not possible to modify its status nor its conclusion further.</p>
</blockquote>
</aside>
<p>But what about when a check run transitions from <code>in_progress</code> to <code>stale</code>? Can I modify the status then?</p>

**Comment by francisfuzz** [comment]
<blockquote>
<p>But what about when a check run transitions from  <code>in_progress</code>  to  <code>stale</code> ? Can I modify the status then?</p>
</blockquote>
<p><a class="mention" href="/danthedaniel">@danthedaniel</a> - Great follow-up question!</p>
<p>Every check run has a <code>status</code> and a <code>conclusion</code>.</p>
<p>A <code>status</code> can be one of the following values: <code>queued</code>, <code>in_progress</code>, or <code>completed</code>. Unless otherwise specified, <code>queued</code> is its default status.</p>
<p>Here’s the order in which a <code>status</code> value can change:</p>
<ul>
<li>
<code>queued</code> :arrow_right: <code>in_progress</code> :arrow_right: <code>completed</code>
</li>
</ul>
<p>If a check run’s <code>status</code> has a value of either <code>queued</code> or <code>in_progress</code> respectively, its <code>conclusion</code> value will be <code>null</code>.</p>
<p>Here’s why: it’s only when a check run’s <code>status</code> is updated to <code>completed</code> that a <code>conclusion</code> must be provided.</p>
<p>The value of <code>conclusion</code> can be one of the following values:</p>
<ul>
<li><code>success</code></li>
<li><code>failure</code></li>
<li><code>neutral</code></li>
<li><code>cancelled</code></li>
<li><code>skipped</code></li>
<li><code>timed_out</code></li>
<li><code>action_required</code></li>
<li>
<code>stale</code> (:warning: only GitHub can update a check run’s <code>conclusion</code> to this value.)</li>
</ul>
<p>Thus, when a check run’s <code>status</code>’ value is <code>completed</code> and its <code>conclusion</code>'s value is <code>stale</code>, it’s not possible to <a href="https://docs.github.com/en/rest/reference/checks#update-a-check-run">update that check run</a> with a different <code>conclusion</code>.</p>
<p>If you have a specific use case for which you would find it helpful to update a check run’s <code>conclusion</code> after the conclusion has been determined, we encourage you to share that with our product team <a href="https://support.github.com/contact/feedback">our official product feedback form</a>. That’s the best place to share requests like these in consideration for future iterations of GitHub features.</p>

**Comment by danthedaniel** [comment]
<p>Thanks for explaining!</p>

**Comment by danthedaniel** [comment]
<p>Actually I have another follow up question. If a run is marked stale can I create a new check run (from the same app) on the same commit as the original check run?</p>
<p>Basically what I want is some way to be able to require a check run from my GitHub app even though it could take &gt; 2 weeks to perform. The tool I’m planning to build would block merging a PR until a certain label is applied to all associated issues (if there’s another way to do this please let me know).</p>

**Comment by francisfuzz** [comment]
<aside class="quote no-group" data-username="danthedaniel" data-post="6" data-topic="120566">
<div class="title">
<div class="quote-controls"></div>
 danthedaniel:</div>
<blockquote>
<p>Actually I have another follow up question. If a run is marked stale can I create a new check run (from the same app) on the same commit as the original check run?</p>
</blockquote>
</aside>
<p>Because a check suite can have more than one check run, you <em>should</em> be able to create as many check runs from the same application as you’d like on the same commit. I’d be curious to see if that <em>isn’t</em> the case!</p>
<aside class="quote no-group" data-username="danthedaniel" data-post="6" data-topic="120566">
<div class="title">
<div class="quote-controls"></div>
 danthedaniel:</div>
<blockquote>
<p>Basically what I want is some way to be able to require a check run from my GitHub app even though it could take &gt; 2 weeks to perform. The tool I’m planning to build would block merging a PR until a certain label is applied to all associated issues (if there’s another way to do this please let me know).</p>
</blockquote>
</aside>
<p>Ahh, thanks for sharing that context! One of my colleagues developed a GitHub App called <code>Semantic Pull Request</code> that might be of interest:</p>
<aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/zeke/semantic-pull-requests" target="_blank">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/15894826/181086604-3fba6eda-f12a-4cb0-bfe5-b98daaf492cb.jpeg" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://github.com/zeke/semantic-pull-requests" target="_blank">zeke/semantic-pull-requests</a></h3>

<p>:robot: Let the robots take care of the semantic versioning - zeke/semantic-pull-requests</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Its sole purpose is to ensure that pull requests follow the <code>Conventional Commits</code> spec. It’s built on <a href="https://probot.github.io/">Probot</a>, a framework for building GitHub Apps in Node.js. You could check out its implementation and possibly adapt it for your project.</p>
<p>I thought it also might be worth sharing––on a cursory search of <a href="https://probot.github.io/apps/">Probot’s List of Apps</a>, I noticed that there’s a <a href="https://probot.github.io/apps/pr-label-enforcer/">pr_label_enforcer</a> app that enforces certain labels on a PR before merging by using status checks (source @ <a href="https://github.com/fossapps/pr_label_enforcer">https://github.com/fossapps/pr_label_enforcer</a>). I can’t say that I’ve used that particular application nor am I able to provide support for either (since I’m neither a maintainer nor owner), but have used <code>semantic-pull-requests</code> in other projects and it works like a charm. :ok_hand:</p>
<p>I hope this helps as some helpful resources to check out as you go about with your project! :raised_hands:</p>

**Comment by danthedaniel** [comment]
<p>Thank you so much! I should have searched around more because PR Label Enforcer looks like just what I need.</p>

**Comment by BenEmdon** [comment, marked as answer]
<blockquote>
<p>But what about when a check run transitions from  <code>in_progress</code>  to  <code>stale</code> ? Can I modify the status then?</p>
</blockquote>
<p>Yes, you can modify a check run which is <code>stale</code> to any other completed state.<br>
Stale conclusions can be updated with the following conclusions:</p>
<ul>
<li>
<code>stale</code> :arrow_right: <code>success</code>
</li>
<li>
<code>stale</code> :arrow_right: <code>failure</code>
</li>
<li>
<code>stale</code> :arrow_right: <code>neutral</code>
</li>
<li>
<code>stale</code> :arrow_right: <code>cancelled</code>
</li>
<li>
<code>stale</code> :arrow_right: <code>skipped</code>
</li>
<li>
<code>stale</code> :arrow_right: <code>timed_out</code>
</li>
<li>
<code>stale</code> :arrow_right: <code>action_required</code>
</li>
</ul>
<p>To update a check run please see our docs on how to <a href="https://docs.github.com/en/rest/reference/checks#update-a-check-run">update a check run</a>.</p>
<p>If you have any feedback about the docs, or this API please share it with our product team on <a href="https://support.github.com/contact/feedback">our official product feedback form</a>.</p>
